### ISSUE TITLE
Conditional email verification

### ISSUE THREAD

**Issue by Samigos** [parse-community/parse-server#7144]
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation

There are occasions where there are multiple groups of users in an app and email verification is needed only to some of them! Right now, Parse Server offers an option of enabling email verification upon signup, but it becomes unconditional... it either works for everyone, or it does not at all.
### Feature / Enhancement Description

To address this limitation, I think there could be a `Boolean` parameter at the `signUp` function that indicates whether email verification is required, or not. 

Also, see initial [forum post](https://community.parseplatform.org/t/conditional-email-verification/1262).

**Comment by mtrezza** [comment]
Thangs for suggesting.

I added the link to your forum post for reference.

> To address this limitation, I think there could be a Boolean parameter at the signUp function that indicates whether email verification is required, or not.

There may also be other places where a change is necessary, for example on login.
Do you want to look into the email verification mechanism in Parse Server and suggest how this could be done?

**Comment by erencihangir** [comment]
HTC cihazımdan gönderildi

----- Yanıt iletisi -----
Gönderen: "Samigos" <notifications@github.com>
Kime: "parse-community/parse-server" <parse-server@noreply.github.com>
Bilgi: "Subscribed" <subscribed@noreply.github.com>
Konu: [parse-community/parse-server] Conditional email verification (#7144)
Tarih: Sal, Oca 26, 2021 19:26

**Comment by mtrezza** [comment]
Possible approaches to discuss:
- a) Add a parameter to the `beforeSave` / `beforeSignup` trigger that includes a parameter to determine there whether a user requires email verification.
- b) Extend the currently existing Parse Server option `verifyUserEmails` to also accept a function (instead of only a boolean value); if a function is set, use the return value of that function (true/ false) to determine per-user whether email verification is needed; the function should provide the full request object, similar to what a trigger would provide.


To extend the existing `_User.emailVerified` field, we could adopt the following approach:
  - field is `undefined`: email doesn't require verification; this is the same value as when the email verification requirement is turned off in the Parse Server option
  - field is `false`: email requires verification but isn't verified yet (or should it always check the function to determine whether verification is needed?, this makes the condition dynamic beyond an initial determination)
  - field is `true`: email is verified

**Comment by mtrezza** [comment]
@mstniy What do you think about option (a), would that solve https://github.com/parse-community/parse-server/issues/8230?

If that works for you, we can look into how to implement it. If the suggestion above to extend the existing `_User.emailVerified` field may be a breaking change, we can look for a different implementation.

**Comment by mstniy** [comment]
@mtrezza Could you elaborate (a) further?
The approach we have implemented in #8231 effectively adds a parameter to the signUp function, after all.

**Comment by mtrezza** [comment]
> adds a parameter to the signUp function

That has been [suggested](https://github.com/parse-community/parse-server/issues/7144#issue-794418982) by @Samigos and that is not an approach we'd pursue. It allows the client to determine whether an email needs to be verified. Only the server should have that authority. If you want to let the client decide you could use the `context` container so tell the server how it should behave. By doing so you of course accept the risk that a user may manipulate that option beyond your control.

> Could you elaborate (a) further?

You can currently use the `beforeSave` and `beforeLogin` Cloud Code trigger to intercept the process when a user tries sign up for a new account or log into an existing account. In that trigger, we could add a parameter that you could manipulate to determine whether the user requires email verification.

For example:

```js
Parse.Cloud.beforeSave(Parse.User, async req => {
  console.log(req.requireEmailVerification); // prints "true" or "false" depending on the Parse Server option setting
  req.requireEmailVerification = true; // override Parse Server option and require email verification for the user
  req.requireEmailVerification = false; // override Parse Server option and don't require email verification for the user
});
```

**Comment by mstniy** [comment]
>It allows the client to determine whether an email needs to be verified. Only the server should have that authority.

The changes in the PR ignore the value of `emailVerified` unless the requester is the master key. Please refer to [the PR description](https://github.com/parse-community/parse-server/pull/8231), as well as [the new test](https://github.com/parse-community/parse-server/pull/8231/files#diff-63c8bb528bb0b39d381a80720d7af56a6ae619273f82c8d6d96ba7c25064d657R3466) and [the relevant lines of code](https://github.com/parse-community/parse-server/pull/8231/files#diff-b15a9afe8c488204e337898e1324ed4b19d8e5154bfa640b9d5e1cc74742b8ceR736)

**Comment by mtrezza** [comment]
Equipping a client with the master key is something we do not advice, so this becomes a non-feature for most applications. Since we don't want to provide the same feature in multiple ways, we'd look for a solution that is universal and usable for most applications. I've added an [example](https://github.com/parse-community/parse-server/issues/7144#issuecomment-1281497430) to maybe give a better idea of (a). Do you think that would cover your required feature?

**Comment by mstniy** [comment]
The clients do not have the master key, so they won't have the option to bypass the verification.
Business logic running as a cloud function, or in another trusted application, can use the master key to bypass the verification.

I don't immediately see how (a) would be useful to us, as the information about whether the email verification should be bypassed or not would not be available to the trigger, but to the party issuing the request (with the master key).

**Comment by mtrezza** [comment]
I had another look at your [PR](https://github.com/parse-community/parse-server/pull/8231). If I understand it correctly, you want to set the user email to be verified, without the user verifying it. But in the [issue](https://github.com/parse-community/parse-server/issues/8230#issue-1407849627) you also write:

> We would like the master key to be able to bypass email verification, suppress the sending of the verification email

That is something that seems to be related to this feature here. With (a) you would disable the email verification for the user (i.e. suppress sending of the verification email), then in Cloud Code you would use the master key to set `emailVerified: true` as if the user had actually verified the email.

So the scope of #8230 is actually 2 functionalities:
- suppress sending of email (this issue here)
- manipulate `emailVerified` field (not within the scope of this issue here)

Is that correct?

**Comment by mstniy** [comment]
True. In #8231, these two functionalities are atomic with respect to each other. Using the masterKey to set `emailVerified` along with signups/email changes suppresses email verification, in addition to setting the value of the field.

**Comment by mtrezza** [comment]
I think we'd want to split this into 2 separate features, it seems each of them have valid use cases on their own, so they shouldn't be depending on each other. Can we address them in 2 separate PRs?

**Comment by mstniy** [comment]
The master key can already set the `emailVerified` field, however, it lacks the ability to bypass email verification for signups/email changes.
Separating these "features" would necessitate the use of some other mechanism to communicate to Parse server that email verification should be bypassed. We think it is intuitive that explicitly setting emailVerified along with a signup/email change bypasses verification. What mechanism would you suggest?

**Comment by mtrezza** [comment]
The [context](https://github.com/parse-community/Parse-SDK-JS/pull/1527) object in the signup method. Whatever you add to the `context` object however must not change the server behavior automatically. It will only be available in Cloud Code to be used in a trigger. So no change in the Parse JSD SDK. The change in Parse Server would be similar to setting the user email verified:

```js
Parse.Cloud.beforeSave(Parse.User, async req => {
  console.log(req.context.sendEmailVerificationEmail); // prints "true" or "false" depending what is passed in `Parse.User.signUp(context)`
  console.log(req.requireEmailVerification); // prints "true" or "false" depending on the Parse Server option setting

  // Override Parse Server behavior
  req.setEmailVerified = true; // sets the email to verified for the user
  req.setEmailVerified = false; // sets the email to unverified for the user
  req.setSendEmailVerificationEmail = true; // Sends the email verification email
  req.setSendEmailVerificationEmail = false; // Doesn't send the email verification email
});
```

Maybe `setEmailVerified` should be set via the master key just like any other property, and only the `setSendEmailVerificationEmail` needs a special solution.

Such a concept allows for example to set the Parse Server option to not require email verification for users by default. For a certain user (e.g. with suspicious activity) email verification could be required.

**Comment by mstniy** [comment]
>  It will only be available in Cloud Code to be used in a trigger. So no change in the Parse JSD SDK. 

I don't see why the master key should not be able to bypass email verification. Sounds like a step backwards to me compared to #8231.

> Maybe setEmailVerified should be set via the master key just like any other property, and only the setSendEmailVerificationEmail needs a special solution.

My point exactly. The master key can already set `emailVerified`.

> Such a concept allows for example to set the Parse Server option to not require email verification for users by default. For a certain user (e.g. with suspicious activity) email verification could be required.

That's a good point. The server deciding, during the signup process, whether the user should be verified certainly is a different use case. This can be implemented using #8231, but it is somewhat tricky: One would need to disallow direct signups by removing public write permission from the _User CLP and create a cloud function that creates signup requests using the master key with the `emailVerified` field set appropriately, in addition to setting the server to `verifyUserEmails: true, preventLoginWithUnverifiedEmail: true`.

Or we could modify #8231 such that `emailVerified` updates are rejected for non-masterkey clients, but is permitted to be set (with its verification-bypassing sideffect) for the master key and any `before*` trigger. That would cover both use cases while keeping the original Parse signup API intact.

**Comment by mstniy** [comment]
Some insight on the last point in my previous message for future reference: RestWrite.js runs beforeSave triggers first, and then calls `transformUser()`, which verifies that either the caller is the master key or the request does not change `emailVerified`. This would prevent a beforeSave trigger from conditionally bypassing email verification. The order of these operations should be changed such that first `emailVerified` is checked to disallow clients from setting it, followed by running the beforeSave triggers and finally sending the verification email or not, depending on whether the request sets `emailVerified`.

**Comment by mtrezza** [comment]
> I don't see why the master key should not be able to bypass email verification.

Yes, I agree that the master key should be able to do that just like with any other non-internal field. Maybe your PR can focus on that part only.

> The server deciding, during the signup process, whether the user should be verified certainly is a different use case. 

As I said, we don't encourage (even though you may have a use case for it), to use the master key on the client side. So any feature suggestion that requires the master key to be used on the client side is considered a "non-feature", or at most it can be a supplemental functionality, but the main feature must be usable without exposing the master key to a client.

Good that you've already studied the code for possible implementations. On a more conceptual level, it seems we are talking about separate functionalities:

- a) Requiring or omitting email verification on a per-user basis, regardless of Parse Server option
- b) Manually setting the `_User.emailVerified` field, just like any other non-internal field
- c) (Re)-sending or suppressing the verification email, so that a developer can design their own email verification flow

I'm still not entirely sure from https://github.com/parse-community/parse-server/pull/8231 how your process would be, but it may only require (a) and not even (b).

1. User A signs up with `userA@example.com` and verifies their email address
2. User B tries to sign up with `userB@example.com`
3. Sever runs Cloud Code trigger `beforeSave(_User)` , finds another user already registered with `example.com` and `_User.emailVerified: true` (genuinely verified), therefore doesn't send a verification email
4. User B tries to change email address
5. Sever runs Cloud Code trigger `beforeSave(_User)`, finds another user already registered with `example.com` and `_User.emailVerified: true` (genuinely verified), therefore doesn't send a verification email

The benefit of not using (b) is that you don't "fake-verify" the user's email address, which would pollute your data and there would be no way of telling which verification is genuine. If at a later point you decide to require verification also from that user, you would simply change the trigger logic, without having to deal with polluted data.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.3.0-alpha.2](https://github.com/parse-community/parse-server/releases/tag/6.3.0-alpha.2)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.4.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.4.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.4.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/6.4.0-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.4.0](https://github.com/parse-community/parse-server/releases/tag/6.4.0)
